src/compiler/checker.ts

```diff
@@ -24122,12 +24122,13 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 if (isIgnoredJsxProperty(source, prop)) {
                     continue;
                 }
-                if (isApplicableIndexType(getLiteralTypeFromProperty(prop, TypeFlags.StringOrNumberLiteralOrUnique), keyType)) {
+                const propKeyType = getLiteralTypeFromProperty(prop, TypeFlags.StringOrNumberLiteralOrUnique);
+                if (isApplicableIndexType(propKeyType, keyType)) {
                     const propType = getNonMissingTypeOfSymbol(prop);
                     const type = exactOptionalPropertyTypes || propType.flags & TypeFlags.Undefined || keyType === numberType || !(prop.flags & SymbolFlags.Optional)
                         ? propType
                         : getTypeWithFacts(propType, TypeFacts.NEUndefined);
-                    const related = isRelatedTo(type, targetInfo.type, RecursionFlags.Both, reportErrors, /*headMessage*/ undefined, intersectionState);
+                    const related = isRelatedTo(type, targetInfo.type, RecursionFlags.Both, reportErrors, /*headMessage*/ undefined, intersectionState, targetInfo === getApplicableIndexInfo(target, propKeyType));

```

Comment:
This is where the `allowExtraPropertyChecks` is toggled~ when relating to a `targetInfo.type`. The source is related to **all** target infos (that's what the `indexSignaturesRelatedTo` caller does). Those checks should only be performed when relating to the applicable index info to reflect what happens in assignments like:
```ts
obj.indexSignatureProp = { someProp: 'foo' }
```

In the example above, `{ someProp: string }` is only related to the applicable index info "selected" by `obj.indexSignatureProp`